[SYCL][RFC]Lowering SYCL Kernel (#11435)
intel/llvm: clang/docs/SYCLFERFCLoweringKernel.md

> # SYCL Kernel

Reviewer: ```suggestion
Discourse topic details:
- Category: [Clang Frontend](https://discourse.llvm.org/c/clang/6)
- Title: "RFC: SYCL kernel lowering"
- Tags: sycl

# SYCL Kernel
```

> A SYCL construct such as `parallel_for` or 'single_task' takes a named function object or a lambda as one of its arguments.   The contents of this function object are executed on the device.   However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, it needs to respect the calling convention of these API.  To enable this, the function object is converted into the format of an OpenCL kernel.

Reviewer 2: ```suggestion
A SYCL construct such as `parallel_for` or `single_task` takes a named function object or a lambda as one of its arguments.   The contents of this function object are executed on the device.   However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, it needs to respect the calling convention of these API.  To enable this, the function object is converted into the format of an OpenCL kernel.
```

There are other cases where you have forward quotes rather than backwards ones (in this RFC and some of the others).  Read through the HTML render, and they are easier to spot.

> A SYCL construct such as `parallel_for` or 'single_task' takes a named function object or a lambda as one of its arguments.   The contents of this function object are executed on the device.   However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, it needs to respect the calling convention of these API.  To enable this, the function object is converted into the format of an OpenCL kernel.

Reviewer: ```suggestion
A SYCL construct such as `parallel_for` or `single_task` takes a named function object or a lambda as one of its arguments.   The contents of this function object are executed on the device.   However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, it needs to respect the calling convention of these API.  To enable this, the function object is converted into the format of an OpenCL kernel.
```

Consider the following code snippet:

```
  sycl::accessor<char, 1, sycl::access::mode::read> acc1;
  int i = 13;
  sycl::sampler smplr;

  struct S {
    char c;
    int i;
  } test_s;

  test_s.c = 14;

  kernel_single_task<class first_kernel>([=]() {
    if (i == 13 && test_s.c == 14) {

      acc1.use();
      smplr.use();
    }
  });

> ```

Reviewer 2: Same comment here about using real SYCL code.  I think you can reuse the same example code I suggested in #11431.

> In this example, the lambda passed to the 'kernel_single_task' construct needs to be executed on the device.  To do this, we propose generating a function object representing the kernel.   The function call operator of this object has the contents of the kernel invocation.  The function object generated looks like:

Author: Thanks, I rephrased it a bit and then am relying on the brief description below.

The device compiler transforms this into (pseudo-code):

```
    spir_kernel void Caller(
       int i,
       struct S test_s,
       __global int* accData, // arg1 of accessor init function
       range<1> accR1,        // arg2 of accessor init function
       range<1> accR2,        // arg3 of accessor init function
       id<1> accId,           // arg4 of accessor init function
       sampler_t smpData      // arg1 of sampler init function
    )
    {
        // Local capture object
        struct FuncObj local;

        // Reassemble capture object from parts
        local.i = i;
        local.s = s;
        // Call acc1 accessor's init function
        sycl::accessor::init(&local.acc1, accData, accR1, accR2, accId);
        // Call smplr sampler's init function
        sycl::accessor::init(&local.smplr, smpData);

        // Call the kernel body
        Callee(&local);
    }

    spir_func void Callee(struct FuncObj* this)
    {
        // body of the kernel invocation
    }
```

The SYCL specification defines rules for allowable types for a kernel parameter.

> The proposed implementation passes the copyable types to the device as separate parameters.  The current implementation is aware of some types such as sycl::accessor and sycl::sampler, for example, which cannot be simply copied from host to device.  (The specification permits this to account for difference in host/device layouts, absence of some fields on either the host or the device, or to allow conversion of pointer values for correct behavior.)  To enable all of this, the parameters of the sycl::accessor and sycl::sampler init functions are transfered from host to device separately.  The values received on the device are passed to the init functions executed on the device, which results in the reassembly of the SYCL object in a form usable on the device.  Note that when such types are elements of an array or a field of a struct or both, special traversal is necessary to pass the type properly.  The proposed mechanism accounts for handling these special instances.

Reviewer 3: >> To enable all of this, the parameters of the sycl::accessor and sycl::sampler init functions are transfered from host to device separately.
As well as streams and annotated_arg (although I don't know if we're upstreaming annotated_arg). Maybe we should say something like "To enable this, such types have an __init function. The parameters of the init functions are transfered from host to device separately"...

Author: I didn't know where to draw the line.   I wanted to add a sampler to show that accessor wasn't the only special type, but didn't want to list stream also.   (annotared_arg totally escaped me.)

# SYCL Kernel

> A SYCL construct such as 'single_task' takes a function object as one of its arguments.   The contents of this function object are executed on the device.   To enable this, the function object is converted into the format of an OpenCL kernel.

Reviewer 4: - People usually knows about `parallel_for` as it is present on any SYCL presentation slides, but `single_task` not so much.
- As this is a FE RFC only, it is not immediately obvious why this transformation is required

```suggestion
A SYCL construct such as `parallel_for` or `single_task` takes a function object as one of its arguments.
The contents of this function object are executed on the device.
However, as the SYCL runtime can rely on other offload APIs like OpenCL or CUDA to execute the function object, therefore it needs to respect the calling convention of these API.
To enable this, the function object is converted into the format of an OpenCL kernel.
```

Consider the following code snippet:

```
  sycl::accessor<char, 1, sycl::access::mode::read> acc1;
  int i = 13;
  sycl::sampler smplr;

  struct S {
    char c;
    int i;
  } test_s;

  test_s.c = 14;

  kernel_single_task<class first_kernel>([=]() {
    if (i == 13 && test_s.c == 14) {

      acc1.use();
      smplr.use();
    }
  });
```

> Here the lambda within the 'kernel_single_task' construct needs to be executed on the device.  To do this, a function object representing the kernel is generated.   The function call operator of this object has the contents of the kernel invocation.  The function object generated looks like:

Reviewer 4: nitpicking

```suggestion
Here the lambda passed to the 'kernel_single_task' construct needs to be executed on the device.  To do this, a function object representing the kernel is generated.   The function call operator of this object has the contents of the kernel invocation.  The function object generated looks like:
```

```
struct FuncObj {
  int i;
  struct S test_s;
  sycl::accessor acc1;
  sycl::sampler smplr;

>   () {  // Function call operator

Reviewer 4: nitpicking again (feels more natural)

```suggestion
  void operator() () {  // Function call operator
```

    if (i == 13 && test_s.c == 14) {
      acc1.use();
      smplr.use();
    }
  }
};
```

> The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.

Reviewer 4: Might be good to expand a bit on the process, from experience many get confused by it.

```suggestion
The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.
It does so by walking the function object data member and producing a parameter for each of them.
Inside the OpenCL kernel, the function object is rebuilt and then called.
Some special types like accessor and sampler are treated a bit differently (see below). 
```

Consider the following code snippet:

```
#include <sycl.hpp>

int main() {
  sycl::queue q;
  sycl::buffer<char> b{sycl::range{1024}};

  q.submit([&](sycl::handler &cgh) {
    sycl::accessor acc{b, cgh};
    int i;
    struct S {
      char c;
      int i;
    } test_s;
    test_s.c = 14;

    cgh.single_task([=] {
      if (i == 13 && test_s.c == 14) {
        acc[0] = 'a';
      }
    });
  });
}
```

In this example, the lambda passed to the `single_task` construct needs to be executed on the device.  The corresponding function object looks like:

```
struct FuncObj {
  int i;
  struct S test_s;
  sycl::accessor acc1;

  void operator () {  // Function call operator
    if (i == 13 && test_s.c == 14) {
      acc[0] = 'a';
    }
  }
};
```

The device compiler then generates a caller in the form of an OpenCL kernel function that calls this function object.  It does so by walking the function object data member and generating a parameter for each of them.  Some special types like accessor are treated a bit differently (see below).  Inside the OpenCL kernel, the function object is rebuilt and then called.

The device compiler transforms this into (pseudo-code):

```

>     spir_kernel void Caller(

Reviewer: Can we avoid references to SPIR or SPIR-V so as to avoid any appearance of a bias?

    }

    spir_func void Callee(struct FuncObj* this)
    {
        // body of the kernel invocation
    }
```

The SYCL specification defines rules for allowable types for a kernel parameter.

The proposed implementation passes the copyable types to the device as separate parameters.  The current implementation is aware of some types such as sycl::accessor, for example, which cannot be simply copied from host to device.  (The specification permits this to account for difference in host/device layouts, absence of some fields on either the host or the device, or to allow conversion of pointer values for correct behavior.)  To enable all of this, these special types have an __init function.  The parameters of this function are transfered from host to device separately.  The values received on the device are passed to the init functions executed on the device, which results in the reassembly of the SYCL object in a form usable on the device.  Note that when such types are elements of an array or a field of a struct or both, special traversal is necessary to pass the type properly.  The proposed mechanism accounts for handling these special instances.

# Location of this logic

> Currently in our implementation, this logic is located in the Sema phase.  Similar to what we are considering for generating the Integration Headers and Foooters, we have an open question between two options that we are considering - one, move this to the CodeGen phase (e.g., OpenMP does their equivalent transformation in CodeGen) and two, move it out of the clang FE and do it in an LLVM IR pass.

Reviewer 3: typo - footers*

Actually it should be singular also - Header and Footer - since we don't have multiple headers/footers
